kube/src/client/mod.rs

```diff
@@ -424,3 +424,70 @@ mod test {
         assert_eq!(s2.details.unwrap().name, ""); // optional probably better..
     }
 }
+
+#[cfg(feature = "ws")]
+// Verify upgrade response according to RFC6455.
+// Based on `tungstenite` and added subprotocol verification.
+fn verify_upgrade_response(res: &Response<Body>, key: &str) -> Result<()> {
+    if res.status() != StatusCode::SWITCHING_PROTOCOLS {
+        return Err(Error::ProtocolSwitch(res.status()));
+    }
+
+    let headers = res.headers();
+    if !headers
+        .get(http::header::UPGRADE)
+        .and_then(|h| h.to_str().ok())
+        .map(|h| h.eq_ignore_ascii_case("websocket"))
+        .unwrap_or(false)
+    {
+        return Err(Error::MissingUpgradeWebSocketHeader);
+    }
+
+    if !headers
+        .get(http::header::CONNECTION)
+        .and_then(|h| h.to_str().ok())
+        .map(|h| h.eq_ignore_ascii_case("Upgrade"))
+        .unwrap_or(false)
+    {
+        return Err(Error::MissingConnectionUpgradeHeader);
+    }
+
+    let accept_key = derive_accept_key(key.as_ref());
+    if !headers
+        .get(http::header::SEC_WEBSOCKET_ACCEPT)
+        .map(|h| h == &accept_key)
+        .unwrap_or(false)
+    {
+        return Err(Error::SecWebSocketAcceptKeyMismatch);
+    }
+
+    // Make sure that the server returned the correct subprotocol.
+    if !headers
+        .get(http::header::SEC_WEBSOCKET_PROTOCOL)
+        .map(|h| h == WS_PROTOCOL)
+        .unwrap_or(false)
+    {
+        return Err(Error::SecWebSocketProtocolMismatch);
+    }
+
+    Ok(())
+}
+
+/// Generate a random key for the `Sec-WebSocket-Key` header.
+/// This must be nonce consisting of a randomly selected 16-byte value in base64.
+#[cfg(feature = "ws")]
+fn sec_websocket_key() -> String {
+    let r: [u8; 16] = rand::random();
+    base64::encode(&r)
+}
```

Comment:
as best i can tell, this should match the rfc? it just seems to be stupidly specific about base64 padding..

Reply:
Yeah, it needs to be base64 value that's decoded to 16 bytes:
```
The request MUST include a header field with the name
|Sec-WebSocket-Key|.  The value of this header field MUST be a
nonce consisting of a randomly selected 16-byte value that has
been base64-encoded (see Section 4 of [RFC4648]).  The nonce
MUST be selected randomly for each connection.

NOTE: As an example, if the randomly selected value was the
sequence of bytes 0x01 0x02 0x03 0x04 0x05 0x06 0x07 0x08 0x09
0x0a 0x0b 0x0c 0x0d 0x0e 0x0f 0x10, the value of the header
field would be "AQIDBAUGBwgJCgsMDQ4PEC=="
```